webapp/src/js/services/telemetry.js

```diff
@@ -0,0 +1,215 @@
+    };
+
+    var storeIt = function(db, key, value) {
+      return db.post({
+        key: key,
+        value: value,
+        date_recorded: Date.now(),
+      });
+    };
+
+    var submitIfNeeded = function(db) {
+      var monthStart = moment().startOf('month');
+      var dbDate = moment(getLastAggregatedDate());
+      if (dbDate.isBefore(monthStart)) {
+        return aggregate(db).then(function() {
+          return reset(db);
+        });
+      }
+    };
+
+    var convertReduceToKeyValues = function(reduce) {
+      var kv = {};
+      reduce.rows.forEach(function(row) {
+        kv[row.key] = row.value;
+      });
+
+      return kv;
+    };
+
+    var generateAggregateDocId = function(metadata) {
+      return [
+        'telemetry',
+        metadata.year,
+        metadata.month,
+        metadata.user,
+        metadata.deviceId,
+      ].join('-');
+    };
+
+    var generateMetadataSection = function() {
+      var date = moment(getLastAggregatedDate());
+      return {
+        year: date.year(),
+        month: date.month(),
+        user: Session.userCtx().name,
+        deviceId: getUniqueDeviceId(),
+      };
+    };
+
+    var generateDeviceStats = function() {
+      return {
+        userAgent: $window.navigator.userAgent,
+        hardwareConcurrency: $window.navigator.hardwareConcurrency,
+        screen: {
+          width: $window.screen.availWidth,
+          height: $window.screen.availHeight,
+        },
+        // TODO: expose some device information in the medic-android wrapper
+        // and pull it in here. Device memory, phone type, stuff like that
+        //
+        // https://github.com/medic/medic-webapp/issues/4882
+      };
+    };
+
+    // This should never happen (famous last words..), because we should only
+    // generate a new document for every month, which is part of the _id.
+    var storeConflictedAggregate = function(aggregateDoc) {
+      aggregateDoc.metadata.conflicted = true;
+      aggregateDoc._id = [aggregateDoc._id, 'conflicted', Date.now()].join('-');
+      return DB().put(aggregateDoc);
+    };
+
+    var aggregate = function(db) {
+      return $q
+        .all([
+          db.query(
+            {
+              map: function(doc, emit) {
+                emit(doc.key, doc.value);
+              },
+              reduce: '_stats',
+            },
+            {
+              group: true,
+            }
+          ),
+          DB().info(),
+        ])
+        .then(function(qAll) {
+          var reduceResult = qAll[0];
+          var infoResult = qAll[1];
+
+          var aggregateDoc = {
+            type: 'telemetry',
+          };
+
+          aggregateDoc.metrics = convertReduceToKeyValues(reduceResult);
+          aggregateDoc.metadata = generateMetadataSection();
+          aggregateDoc._id = generateAggregateDocId(aggregateDoc.metadata);
+          aggregateDoc.device = generateDeviceStats();
+          aggregateDoc.dbInfo = infoResult;
+
+          return DB()
```

Comment:
For performance reasons would it be better to put the aggregate docs in the individual user's meta DB?  Presumably these docs are write-once and never relevant to other users?  I appreciate that this would add work on the server-side to collect all telemetry docs for different users.

Reply:
Yeah I think that's a great call. It does add more work on the server side, but it's really just a 50-line script someone has to write and a design doc view on the meta DB (is that hard for some reason? Do we already have one?) to pull out `telemetry` docs.

Thoughts @garethbowen?

Reply:
We do already have a ddoc on the meta DB, but you might also be able to just query using all_docs because the doc _id is useful.

But yes I think it makes sense to store it in the user's meta db.

We need to think about how we're going to get this data out at some point...